[lldb] Detect when full DWARF debugging should be enabled

augusto2112 opened the pull request:
Automatically turn on full DWARF debugging if reflection metadata isn't available.

rdar://122711892

augusto2112 commented:
@swift-ci test

adrian-prantl commented on lldb/include/lldb/lldb-private-enumerations.h:
```diff
@@ -172,10 +172,10 @@ typedef enum SwiftModuleLoadingMode {
-  eAutoDetectSwiftCxxInterop,
-  eEnableSwiftCxxInterop,
-  eDisableSwiftCxxInterop
+enum AutoBool {
```
My C++ is lacking here – does this type autoconvert to bool? We want to make sure that `if ([](){ return eAutoBoolFalse;})` does the right thing and is a type check error.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/ReflectionContext.cpp:
```diff
@@ -32,21 +32,21 @@ struct DescriptorFinderForwarder : public swift::reflection::DescriptorFinder {
   std::unique_ptr<swift::reflection::BuiltinTypeDescriptorBase>
   getBuiltinTypeDescriptor(const swift::reflection::TypeRef *TR) override {
-    if (m_descriptor_finder)
+    if (m_descriptor_finder && shouldConsultDescriptorFinder())
```
This pattern looks like it could be factored out into a template or a function that takes a std::function?

augusto2112 replied:
Maybe there's a better way to do this, but I think the "obvious" template solution is more trouble than it's worth. We'd need a templated function like:

```
  template <typename T>
  std::unique_ptr<T>
  getDescriptor(std::function<std::unique_ptr<T>()> descriptor_func) {
    if (m_descriptor_finder && shouldConsultDescriptorFinder())
      return descriptor_func();
    return nullptr;
  }
```

And then on the call side we'd need something like:

```
  std::unique_ptr<swift::reflection::BuiltinTypeDescriptorBase>
  getBuiltinTypeDescriptor(const swift::reflection::TypeRef *TR) override {
    std::function<
        std::unique_ptr<swift::reflection::BuiltinTypeDescriptorBase>()>
        lambda =
            [&]() { return m_descriptor_finder->getBuiltinTypeDescriptor(TR); };
    return getDescriptor(lambda);
  }

```

We need to spell out the lambda's type because the compiler is not able to deduce them.

adrian-prantl commented on lldb/source/Plugins/LanguageRuntime/Swift/ReflectionContext.cpp:
```diff
@@ -57,8 +57,24 @@ struct DescriptorFinderForwarder : public swift::reflection::DescriptorFinder {
+      case lldb_private::eAutoBoolFalse:
+        return false;
+      case lldb_private::eAutoBoolAuto:
+        return !m_image_added;
```
Can you comment how this works here? It's not obvious why `m_image_added` is the condition here.

augusto2112 committed:
[lldb] Detect when full DWARF debugging should be enabled

Automatically turn on full DWARF debugging if reflection metadata isn't available.

rdar://122711892

augusto2112 commented:
@swift-ci test

augusto2112 commented:
@swift-ci test Windows

augusto2112 commented:
@swift-ci test windows
